activitysim/core/util.py

```diff
@@ -371,6 +379,79 @@ def assign_in_place(df, df2):
 
     df[new_columns] = df2[new_columns]
 
+    for c in new_columns:
+        if pd.api.types.is_object_dtype(df[c]):
+            df[c] = df[c].astype("category")
+
+    auto_opt_pd_dtypes(df, downcast_int, downcast_float, inplace=True)
+
+
+def auto_opt_pd_dtypes(
+    df_: pd.DataFrame, downcast_int=False, downcast_float=False, inplace=False
+) -> Optional[pd.DataFrame]:
+    """
+    Automatically downcast Number dtypes for minimal possible,
+    will not touch other (datetime, str, object, etc)
+
+    Parameters
+    ----------
+    df_ : pd.DataFrame
+        assignment left-hand-side (dest)
+    downcast_int: bool
+        if True, downcast int columns if possible
+    downcast_float: bool
+        if True, downcast float columns if possible
+    inplace: bool
+        if False, will return a copy of input dataset
+
+    Returns
+    -------
+        `None` if `inplace=True` or dataframe if `inplace=False`
+
+    """
+    df = df_ if inplace else df_.copy()
+
+    for col in df.columns:
+        dtype = df[col].dtype
+        # Skip optimizing floats for precision concerns
+        if pd.api.types.is_float_dtype(dtype):
+            if not downcast_float:
+                continue
+            else:
+                # there is a bug in pandas to_numeric
+                # when convert int and floats gt 16777216
+                # https://github.com/pandas-dev/pandas/issues/43693
+                # https://github.com/pandas-dev/pandas/issues/23676#issuecomment-438488603
+                if df[col].max() >= 16777216:
+                    continue
+                else:
+                    df[col] = pd.to_numeric(df[col], downcast="float")
+        # Skip if the column is already categorical
+        if pd.api.types.is_categorical_dtype(dtype):
+            continue
+        # Handle integer types
+        if pd.api.types.is_integer_dtype(dtype):
+            if not downcast_int:
+                continue
+            # there is a bug in pandas to_numeric
+            # when convert int and floats gt 16777216
+            # https://github.com/pandas-dev/pandas/issues/43693
+            # https://github.com/pandas-dev/pandas/issues/23676#issuecomment-438488603
+            if df[col].max() >= 16777216:
+                continue
+            else:
+                df[col] = pd.to_numeric(df[col], downcast="integer")
+                continue
+            # there are calcualtions in asim that expect results in negative values
+            # operations on two unsigned types will not produce negative values
+            # if df[col].min() >= 0:
+            #     df[col] = pd.to_numeric(df[col], downcast='unsigned')
+            # else:
+            #     df[col] = pd.to_numeric(df[col], downcast='integer')
```

Comment:
Yep, I deleted them. I should probably open an issue about the negative values (although it's not really an issue), since it impacts on the data type to use. It would be good to remember this if/when we do the data model.